File: ServeSys/ServeSys.API/Modules/Shared/DTOs/ApiResponse.cs
namespace ServeSys.API.Modules.Shared.DTOs;

public class ApiResponse<T>
{
    public bool Success { get; set; }
    public string Message { get; set; } = string.Empty;
    public T? Data { get; set; }
    
    // Chứa danh sách lỗi chi tiết
    public object? Errors { get; set; }

    public static ApiResponse<T> Ok(T? data, string message = "Success")
    {
        return new ApiResponse<T> { Success = true, Message = message, Data = data };
    }

    public static ApiResponse<T> Fail(string message, object? errors = null)
    {
        return new ApiResponse<T> { Success = false, Message = message, Errors = errors };
    }
}

public class ApiResponse
{
    public static ApiResponse<T> Ok<T>(T data, string message = "Success")
       => ApiResponse<T>.Ok(data, message);

    public static ApiResponse<object> Fail(string message, object? errors = null)
        => ApiResponse<object>.Fail(message, errors);
}

File: ServeSys.Mobile/servesys/lib/core/network/api_response.dart
class ApiError {
  final String code;
  final String description;

  const ApiError({required this.code, required this.description});

  factory ApiError.fromJson(Map<String, dynamic> json) {
    return ApiError(
      code: (json['code'] as String?) ?? '',
      description: (json['description'] as String?) ?? '',
    );
  }
}

class ApiResponse<T> {
  final bool success;
  final String? message;
  final T? data;
  final List<ApiError> errors;

  ApiResponse({
    required this.success,
    this.message,
    this.data,
    this.errors = const []
  });

  /// Returns the most user-friendly error string available:
  /// 1. First error description from [errors] (most specific)
  /// 2. [message] from the top-level response
  /// 3. Fallback empty string
  List<ApiError> get allErrorMessages {
    if (errors.isNotEmpty) {
      return errors;
    }
    if (message != null && message!.isNotEmpty) {
      return [ApiError(code: '', description: message!)];
    }
    return [];
  }

  ApiError get firstErrorMessage {
    // Priority 1: errors array (most specific, e.g. field-level validation)
    if (errors.isNotEmpty && errors.first.description.isNotEmpty) {
      return errors.first;
    }
    // Priority 2: top-level message (server returns only message without errors)
    if (message != null && message!.isNotEmpty) {
      return ApiError(code: '', description: message!);
    }
    return const ApiError(code: '', description: '');
  }

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic)? fromJsonT,
  ) {
    final rawErrors = json['errors'];
    final errors = (rawErrors is List)
        ? rawErrors
            .whereType<Map<String, dynamic>>()
            .map(ApiError.fromJson)
            .toList()
        : <ApiError>[];

    return ApiResponse(
      success: json['success'] ?? false,
      message: json['message'],
      data: json['data'] != null && fromJsonT != null
          ? fromJsonT(json['data'])
          : null,
      errors: errors,
    );
  }
}
